### PR TITLE
Bug fixes/improvements around foraging

### PR DESCRIPTION
**Pull request by mpardalos**
+ [x] Addressed #123
+ [X] Used the typedefs for ClientID and Resources wherever necessary

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by lhl2617** [comment]
This was done as a ~pair~team-programming effort by Team1. @QFSW, @Darrekt and I are disqualified from approving this.

File: internal/common/baseclient/baseclient.go [hunk 1]
```diff
@@ -25,23 +25,25 @@ type Client interface {
...
+	GetTaxContribution() shared.Resources
+	RequestAllocation() shared.Resources
 
 	//IIFO: OPTIONAL
 	MakePrediction() (shared.PredictionInfo, error)
 	ReceivePredictions(receivedPredictions shared.PredictionInfoDict) error
 
 	//Foraging
 	DecideForage() (shared.ForageDecision, error)
+	ForageUpdate(forageReturn int)
```
Comment (lhl2617) [inline internal/common/baseclient/baseclient.go]:
```suggestion
	ForageUpdate(forageReturn shared.Resources)
```

File: internal/common/foraging/deerhunt.go [hunk 2]
```diff
@@ -17,40 +17,41 @@ type deerHuntParams struct {
...
 	fConf := config.GameConfig().ForagingConfig
 
 	input := d.TotalInput()
 	decay := fConf.IncrementalInputDecay
 	maxDeer := fConf.MaxDeerPerHunt
 	nDeerFromInput := deerUtilityTier(input, maxDeer, decay) // get max number of deer allowed for given resource input
-	utility := 0.0
-	for i := uint(1); i < nDeerFromInput; i++ {
+	utility := shared.Resources(0.0)
+	for i := uint(0); i < nDeerFromInput; i++ {
```
Comment (lhl2617) [inline internal/common/foraging/deerhunt.go]: @JamesTev can you look into this?

File: internal/server/iigointernal/base_president_test.go [hunk 3]
```diff
@@ -9,42 +9,85 @@ import (
...
 
 func TestAllocationRequests(t *testing.T) {
 	cases := []struct {
 		name  string
 		input allocInput
-		reply map[int]int
+		reply map[shared.ClientID]shared.Resources
 		want  error
 	}{
+		// FIXME: @mpardalos could not get this running
```
Comment (lhl2617) [inline internal/server/iigointernal/base_president_test.go]: @Neelesh99 @preetl fyi

**Comment by lhl2617** [comment]
_silence, linter_

File: internal/server/iigointernal/base_president_test.go (hunk 3 above)
Comment (Neelesh99) [reply internal/server/iigointernal/base_president_test.go]: Had a look, it seems the tests were written with an int version of resources in mind. So the expected results which we put in are effectively a floor function on the actual float numbers. Will fix and request to push

Comment (Neelesh99) [reply internal/server/iigointernal/base_president_test.go]: Ok tests are fixed ;)

**Review by Neelesh99 — APPROVED** [review]
PR looks good, IIGO certainly all ok. The foraging changes seem fine on surface @JamesTev might want to take a look and confirm. Thanks for the Resources type ;)

**Comment by mpardalos** [comment]
I reverted that off-by-1 change, so James can just have a look and make that change later. This should be good to merge now.

**Comment by lhl2617** [comment]
image

**Pull request merged**